libafl/src/executors/inprocess/with_state.rs

```diff
@@ -0,0 +1,416 @@
+use alloc::boxed::Box;
+use core::{
+    borrow::BorrowMut,
+    fmt::{self, Debug, Formatter},
+    marker::PhantomData,
+    time::Duration,
+};
+
+use libafl_bolts::tuples::tuple_list;
+
+use crate::{
+    events::{EventFirer, EventRestarter},
+    executors::{
+        hooks::{inprocess::InProcessHooks, ExecutorHooksTuple},
+        inprocess::{GenericInProcessExecutorInner, HasInProcessHooks},
+        Executor, ExitKind, HasObservers,
+    },
+    feedbacks::Feedback,
+    fuzzer::HasObjective,
+    inputs::UsesInput,
+    observers::{ObserversTuple, UsesObservers},
+    state::{HasCorpus, HasExecutions, HasSolutions, State, UsesState},
+    Error,
+};
+
+/// The process executor simply calls a target function, as mutable reference to a closure
+/// The internal state of the executor is made available to the harness.
+pub type InProcessExecutorWithState<'a, H, OT, S, ES> =
+    GenericInProcessExecutorWithState<H, &'a mut H, (), OT, S, ES>;
+
+/// The process executor simply calls a target function, as boxed `FnMut` trait object
+/// The internal state of the executor is made available to the harness.
+pub type OwnedInProcessExecutor<OT, S, ES> = GenericInProcessExecutorWithState<
+    dyn FnMut(&<S as UsesInput>::Input, &mut ES) -> ExitKind,
+    Box<dyn FnMut(&<S as UsesInput>::Input, &mut ES) -> ExitKind>,
+    (),
+    OT,
+    S,
+    ES,
+>;
+
+/// The inmem executor simply calls a target function, then returns afterwards.
+/// The harness can access the internal state of the executor.
+#[allow(dead_code)]
+pub struct GenericInProcessExecutorWithState<H, HB, HT, OT, S, ES>
+where
+    H: FnMut(&S::Input, &mut ES) -> ExitKind + ?Sized,
+    HB: BorrowMut<H>,
+    HT: ExecutorHooksTuple,
+    OT: ObserversTuple<S>,
+    S: State,
+{
+    /// The harness function, being executed for each fuzzing loop execution
+    harness_fn: HB,
+    /// The state used as argument of the harness
+    pub executor_state: ES,
+    /// Inner state of the executor
```

Comment:
so it's not state it's more 
`Inner executor with state` right?

Reply:
I was thinking of putting it in `inner`, to merge the different states in the same struct.
It would also let me remove this ugly `pub`.